// @threlte/xr/dist/internal/stores.js
import { currentWritable } from '@threlte/core';
export const isPresenting = currentWritable(false);
export const isHandTracking = currentWritable(false);
export const session = currentWritable(undefined);
export const referenceSpaceType = currentWritable(undefined);
export const xr = currentWritable(undefined);
export const controllerDispatchers = {
    left: currentWritable(undefined),
    right: currentWritable(undefined)
};
export const handDispatchers = {
    left: currentWritable(undefined),
    right: currentWritable(undefined)
};
export const teleportState = currentWritable({
    left: {
        enabled: false,
        hovering: false
    },
    right: {
        enabled: false,
        hovering: false
    }
});
export const teleportIntersection = {
    left: currentWritable(undefined),
    right: currentWritable(undefined)
};
export const pointerState = currentWritable({
    left: {
        enabled: false,
        hovering: false
    },
    right: {
        enabled: false,
        hovering: false
    }
});
export const pointerIntersection = {
    left: currentWritable(undefined),
    right: currentWritable(undefined)
};
